File: python/src/linodemcp/audit/reader.py
"""Recent-events reader over the JSONL audit log.

Mirrors ``go/internal/audit/reader.go``. Scans the active ``audit.log``
and rotated ``audit-YYYY-MM-DD.log[.gz]`` files, newest first, applying
optional filters and returning up to a limit.
"""

from __future__ import annotations

import fnmatch
import gzip
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING

from linodemcp.audit.event import Event
from linodemcp.audit.jsonl import ACTIVE_LOG_FILE_NAME
from linodemcp.audit.retention import parse_rotated_file_day

if TYPE_CHECKING:
    from datetime import date, datetime

    from linodemcp.audit.event import Capability, Status

# Default number of events returned when the caller does not specify a
# limit. Mirrors the Go DefaultRecentLimit.
DEFAULT_RECENT_LIMIT = 20

# Hard cap so a single query can't pull an unbounded slice into memory.
MAX_RECENT_LIMIT = 200


@dataclass
class RecentQuery:
    """Filters for a recent-events scan.

    Unset fields (None / empty / 0) mean "no constraint". ``tool`` is a
    glob (fnmatch syntax); ``capability`` and ``status`` are exact
    matches; ``since`` and ``until`` bound the event timestamp
    (inclusive). Meta events are excluded unless ``include_meta``.
    """

    limit: int = 0
    since: datetime | None = None
    until: datetime | None = None
    tool: str = ""
    capability: Capability | str = ""
    status: Status | str = ""
    include_meta: bool = field(default=False)


def read_recent(directory: str, query: RecentQuery) -> list[Event]:
    """Return up to ``limit`` events, newest first, matching the filters.

    Scans the active ``audit.log`` first (today's events), then rotated
    files in date-descending order, stopping once the limit is reached.
    A missing directory returns an empty list (querying before the
    first event is normal). Undecodable lines and unreadable files are
    skipped so one corrupt record doesn't abort the scan.
    """
    limit = query.limit
    if limit <= 0:
        limit = DEFAULT_RECENT_LIMIT
    limit = min(limit, MAX_RECENT_LIMIT)

    base = Path(directory)
    if not base.is_dir():
        return []

    results: list[Event] = []

    for name in _ordered_audit_files(base):
        events = _read_events_from_file(base / name)

        # Within a file lines are append-order (oldest first); walk
        # backwards so the accumulator stays newest-first across the
        # date-descending file order.
        for event in reversed(events):
            if not _matches(query, event):
                continue

            results.append(event)
            if len(results) >= limit:
                return results

    return results


def scan_events(
    directory: str, since: datetime | None, include_meta: bool
) -> list[Event]:
    """Return every event at or after ``since`` (None = all), honoring
    ``include_meta``. Unlike :func:`read_recent` there is no count limit:
    aggregation (Phase 3d summary, 3f export) needs the whole window. A
    missing directory returns an empty list.
    """
    base = Path(directory)
    if not base.is_dir():
        return []

    events: list[Event] = []

    for name in _ordered_audit_files(base):
        for event in _read_events_from_file(base / name):
            if not include_meta and event.tool_capability == "meta":
                continue

            if since is not None and event.ts < since:
                continue

            events.append(event)

    return events


def _matches(query: RecentQuery, event: Event) -> bool:
    """Report whether an event satisfies every set filter."""
    if not query.include_meta and event.tool_capability == "meta":
        return False

    if query.tool and not fnmatch.fnmatchcase(event.tool, query.tool):
        return False

    if query.capability and event.tool_capability != query.capability:
        return False

    if query.status and event.status != query.status:
        return False

    if query.since is not None and event.ts < query.since:
        return False

    return not (query.until is not None and event.ts > query.until)


def _ordered_audit_files(base: Path) -> list[str]:
    """Return audit file names to scan, newest first.

    The active ``audit.log`` (if present) comes first, followed by
    rotated files in date-descending order.
    """
    has_active = False
    rotated: list[tuple[date, str]] = []

    for entry in base.iterdir():
        if not entry.is_file():
            continue

        name = entry.name
        if name == ACTIVE_LOG_FILE_NAME:
            has_active = True
            continue

        day = parse_rotated_file_day(name)
        if day is not None:
            rotated.append((day, name))

    rotated.sort(key=lambda pair: pair[0], reverse=True)

    names: list[str] = []
    if has_active:
        names.append(ACTIVE_LOG_FILE_NAME)

    names.extend(name for _, name in rotated)
    return names


def _read_events_from_file(path: Path) -> list[Event]:
    """Decode every JSON line in ``path`` into an Event, oldest first.

    Gzipped rotated files are decompressed transparently. Open failures
    yield an empty list (the scan skips the file); undecodable lines are
    skipped individually.
    """
    try:
        raw = _read_file_text(path)
    except OSError:
        return []

    events: list[Event] = []

    for line in raw.splitlines():
        stripped = line.strip()
        if not stripped:
            continue

        try:
            events.append(Event.from_dict(json.loads(stripped)))
        except (ValueError, KeyError):
            continue

    return events


def _read_file_text(path: Path) -> str:
    """Read a log file as text, decompressing .gz rotated files."""
    if path.name.endswith(".gz"):
        with gzip.open(path, "rt", encoding="utf-8") as handle:
            return handle.read()

    return path.read_text(encoding="utf-8")

File: go/internal/audit/reader.go
package audit

import (
	"bufio"
	"compress/gzip"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path"
	"slices"
	"sort"
	"time"
)

// DefaultRecentLimit is the default number of events returned by a
// recent-events query when the caller does not specify a limit.
const DefaultRecentLimit = 20

// MaxRecentLimit caps the recent-events query so a single call can't
// pull an unbounded slice into memory. Callers asking for more are
// clamped to this value.
const MaxRecentLimit = 200

// RecentQuery filters a recent-events scan. Zero-valued fields mean
// "no constraint" for that dimension. Tool is a glob (path.Match
// syntax); Capability and Status are exact matches; Since and Until
// bound the event timestamp (inclusive lower, inclusive upper).
type RecentQuery struct {
	Limit       int
	Since       time.Time
	Until       time.Time
	Tool        string
	Capability  Capability
	Status      Status
	IncludeMeta bool
}

// ReadRecent returns up to Limit events from the audit directory,
// newest first, matching the query filters. It scans the active
// audit.log first (today's events), then rotated audit-YYYY-MM-DD.log
// and .log.gz files in date-descending order, stopping once Limit
// matches are collected.
//
// A missing directory (no audit ever written) returns an empty slice,
// not an error: querying before the first event is a normal state.
// Undecodable lines are skipped so a single corrupt record doesn't
// abort the scan. File-open failures mid-scan are skipped too; the
// returned error is reserved for a directory that exists but can't be
// listed.
func ReadRecent(dir string, query *RecentQuery) ([]Event, error) {
	limit := query.Limit
	if limit <= 0 {
		limit = DefaultRecentLimit
	}

	if limit > MaxRecentLimit {
		limit = MaxRecentLimit
	}

	root, err := openReadRoot(dir)
	if err != nil {
		if errors.Is(err, errAuditDirMissing) {
			return []Event{}, nil
		}

		return nil, err
	}

	defer func() { _ = root.Close() }()

	files, err := orderedAuditFiles(root)
	if err != nil {
		return nil, fmt.Errorf("audit: list audit dir %s: %w", dir, err)
	}

	results := make([]Event, 0, limit)

	for _, name := range files {
		events := readEventsFromFile(root, name)

		// Within a file lines are append-order (oldest first); walk
		// backwards so the accumulator stays newest-first across the
		// date-descending file order.
		for i := range slices.Backward(events) {
			if !query.matches(&events[i]) {
				continue
			}

			results = append(results, events[i])
			if len(results) >= limit {
				return results, nil
			}
		}
	}

	return results, nil
}

// matches reports whether an event satisfies every set filter in the
// query. Meta events are excluded unless IncludeMeta is set.
func (q *RecentQuery) matches(event *Event) bool {
	if !q.IncludeMeta && event.ToolCapability == CapabilityMeta {
		return false
	}

	if q.Tool != "" {
		ok, err := path.Match(q.Tool, event.Tool)
		if err != nil || !ok {
			return false
		}
	}

	if q.Capability != "" && event.ToolCapability != q.Capability {
		return false
	}

	if q.Status != "" && event.Status != q.Status {
		return false
	}

	if !q.Since.IsZero() && event.TS.Before(q.Since) {
		return false
	}

	if !q.Until.IsZero() && event.TS.After(q.Until) {
		return false
	}

	return true
}

// openReadRoot opens an os.Root on dir for reading. A missing
// directory returns errAuditDirMissing so callers can treat "queried
// before the first event" as an empty result rather than a hard error
// (and without a (nil, nil) return). Other open failures are wrapped.
func openReadRoot(dir string) (*os.Root, error) {
	root, err := os.OpenRoot(dir)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return nil, errAuditDirMissing
		}

		return nil, fmt.Errorf("audit: open read root %s: %w", dir, err)
	}

	return root, nil
}

// orderedAuditFiles returns the audit file names to scan, newest
// first: the active audit.log (if present) followed by rotated files
// in date-descending order. Caller must hold the root open.
func orderedAuditFiles(root *os.Root) ([]string, error) {
	entries, err := fs.ReadDir(root.FS(), ".")
	if err != nil {
		return nil, fmt.Errorf("read dir: %w", err)
	}

	var (
		hasActive bool
		rotated   []rotatedFile
	)

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}

		name := entry.Name()
		if name == ActiveLogFileName {
			hasActive = true

			continue
		}

		if day, ok := parseRotatedFileDay(name); ok {
			rotated = append(rotated, rotatedFile{name: name, day: day})
		}
	}

	// Newest rotated date first.
	sort.Slice(rotated, func(i, j int) bool {
		return rotated[i].day.After(rotated[j].day)
	})

	names := make([]string, 0, len(rotated)+1)
	if hasActive {
		names = append(names, ActiveLogFileName)
	}

	for _, file := range rotated {
		names = append(names, file.name)
	}

	return names, nil
}

// rotatedFile pairs a rotated file name with its parsed day so the
// scan can sort by date without re-parsing.
type rotatedFile struct {
	name string
	day  time.Time
}

// readEventsFromFile decodes every JSON line in name into an Event,
// in file order (oldest first). Gzipped rotated files are
// decompressed transparently. Returns whatever decoded successfully;
// open failures yield an empty slice (the scan skips the file), and
// undecodable lines are skipped individually.
func readEventsFromFile(root *os.Root, name string) []Event {
	file, err := root.Open(name)
	if err != nil {
		return nil
	}

	defer func() { _ = file.Close() }()

	var reader io.Reader = file

	if isGzipName(name) {
		gzReader, gzErr := gzip.NewReader(file)
		if gzErr != nil {
			return nil
		}

		defer func() { _ = gzReader.Close() }()

		reader = gzReader
	}

	var events []Event

	scanner := bufio.NewScanner(reader)
	scanner.Buffer(make([]byte, 0, bufScanInitial), bufScanMax)

	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}

		var event Event
		if err := json.Unmarshal(line, &event); err != nil {
			continue
		}

		events = append(events, event)
	}

	return events
}

// Scanner buffer sizes. Audit lines are small (a few hundred bytes
// typically), but a deeply nested args map could push past the
// default 64KB token cap, so the max is raised to keep large events
// readable rather than silently truncated.
const (
	bufScanInitial = 64 * 1024
	bufScanMax     = 1024 * 1024
)

// isGzipName reports whether name is a gzipped rotated log.
func isGzipName(name string) bool {
	const gzSuffix = ".gz"

	return len(name) >= len(gzSuffix) && name[len(name)-len(gzSuffix):] == gzSuffix
}
